### PR TITLE
Add finish reason to Generation for usage downstream

### PR DESCRIPTION
**Pull request by creatorrr**
Add `finish_reason` to `Generation` as well as extend `BaseOpenAI._generate` to include it in the output. This can be useful for usage in downstream tasks when we need to filter for only generations that finished because of `"stop"` for example. Maybe we should add this to `LLMChain` as well?

For more details, see https://beta.openai.com/docs/guides/completion/best-practices

### CODE REVIEW

**Comment by creatorrr** [comment]
For instance, I'm using it like so:

```python
class FinishReasonChain(LLMChain):
    prefer_finish_on: Optional[str] = None
    
    def apply(self, input_list: List[Dict[str, Any]]) -> List[Dict[str, str]]:
        """Utilize the LLM generate method for speed gains."""
        response = self.generate(input_list)
        outputs = []
        for generation in response.generations:
            # Get the text of the top generated string.
            response_str = generation[0].text
            
            # If `prefer_finish_on_stop` then try to pick that
            if self.prefer_finish_on:
                responses = sorted(
                    generation,
                    key=lambda g: g.finish_reason == self.prefer_finish_on,
                    reverse=True
                )
                
                response_str = responses[0].text
            
            if self.verbose:
                langchain.logger.log_llm_response(response_str)
            outputs.append({self.output_key: response_str})
        return outputs

```

File: langchain/schema.py [hunk 1]
```diff
@@ -23,4 +23,7 @@ class Generation(NamedTuple):
 
     text: str
     """Generated text output."""
+
+    finish_reason: Optional[str] = None
+    """Reason for finishing (OpenAI specific)"""
```
Comment (hwchase17) [inline langchain/schema.py:28]: if this is openai specific, lets just add a `generation_info` optional dictionary where providers can put provider specific information. that way this doesnt grow in size if we have a bunch of provider specific fields

we do this for LLMOutput: https://github.com/hwchase17/langchain/blob/6d78be0c8341baec4f9c5559db8c85d8956a7a74/langchain/llms/base.py#L20

Comment (creatorrr) [reply langchain/schema.py:28]: Makes sense. lemme update

Comment (creatorrr) [reply langchain/schema.py]: Made changes, lmk what you think

**Review by hwchase17 — APPROVED** [review]
amazing! will merge in to todays release. thank you!

**Comment by creatorrr** [comment]
IMHO, prefer outputs with `finish_reson=='stop'` should be part of the default implementation. OpenAI itself recommends it. Do you think it makes sense to patch the `OpenAIChain` with that. Something like what I mentioned earlier:

```python
class FinishReasonChain(LLMChain):
    prefer_finish_on: Optional[str] = None
    
    def apply(self, input_list: List[Dict[str, Any]]) -> List[Dict[str, str]]:
        """Utilize the LLM generate method for speed gains."""
        response = self.generate(input_list)
        outputs = []
        for generation in response.generations:
            # Get the text of the top generated string.
            response_str = generation[0].text
            
            # If `prefer_finish_on_stop` then try to pick that
            if self.prefer_finish_on:
                responses = sorted(
                    generation,
                    key=lambda g: g.finish_reason == self.prefer_finish_on,
                    reverse=True
                )
                
                response_str = responses[0].text
            
            if self.verbose:
                langchain.logger.log_llm_response(response_str)
            outputs.append({self.output_key: response_str})
        return outputs
```

**Comment by hwchase17** [comment]
> IMHO, prefer outputs with `finish_reson=='stop'` should be part of the default implementation. OpenAI itself recommends it. Do you think it makes sense to patch the `OpenAIChain` with that. Something like what I mentioned earlier:
> 
> ```python
> class FinishReasonChain(LLMChain):
>     prefer_finish_on: Optional[str] = None
>     
>     def apply(self, input_list: List[Dict[str, Any]]) -> List[Dict[str, str]]:
>         """Utilize the LLM generate method for speed gains."""
>         response = self.generate(input_list)
>         outputs = []
>         for generation in response.generations:
>             # Get the text of the top generated string.
>             response_str = generation[0].text
>             
>             # If `prefer_finish_on_stop` then try to pick that
>             if self.prefer_finish_on:
>                 responses = sorted(
>                     generation,
>                     key=lambda g: g.finish_reason == self.prefer_finish_on,
>                     reverse=True
>                 )
>                 
>                 response_str = responses[0].text
>             
>             if self.verbose:
>                 langchain.logger.log_llm_response(response_str)
>             outputs.append({self.output_key: response_str})
>         return outputs
> ```

yes i like that a lot!

**Pull request merged**